### PR TITLE
Bitcoin Support

### PR DESCRIPTION
**Pull request by sisou**

### CODE REVIEW

**Review by Yukiioz — APPROVED** [review]
Didn't found any real issue. Only minor comments.
Well done.

File: src/views/ActivateBitcoinSuccess.vue [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        this.walletInfo = walletInfo;
+
+        const btcAddresses = await WalletInfoCollector.deriveBitcoinAddresses(this.keyguardResult.bitcoinXPub);
+
+        this.walletInfo.btcXPub = this.keyguardResult.bitcoinXPub;
+        this.walletInfo.btcAddresses = btcAddresses;
+
+        WalletStore.Instance.put(this.walletInfo);
+
+        // if (this.receiptsError) {
```
[the remark this answers is not in the document]
Comment (danimoh) [reply src/views/ActivateBitcoinSuccess.vue:59]: Note: I implemented error handling for all electrum / btc requests, but it's on top of the swap branch

File: demos/Demo.ts [hunk 2]
```diff
@@ -399,6 +411,56 @@ class Demo {
...
+                throw new Error('Cannot use legacy account');
+            }
+
+            const senderAddress = account.btcAddresses ? account.btcAddresses.external[0] : null;
+            if (!senderAddress) {
+                alert('No BTC address found in account, activate Bitcoin for this account first');
+                throw new Error('No BTC address found');
+            }
+
+            const txRequest: SignBtcTransactionRequest = {
```
[the remark this answers is not in the document]
Comment (sisou) [reply demos/Demo.ts:434]: Fixed in #431: https://github.com/nimiq/hub/commit/918f63cf0d9ddedb789ca01ad4e1f11ddcf6189b

File: package.json [hunk 3]
```diff
@@ -18,8 +20,9 @@
     },
     "dependencies": {
         "@nimiq/browser-warning": "^1.1.0",
+        "@nimiq/electrum-client": "https://github.com/nimiq/electrum-client#build",
         "@nimiq/iqons": "^1.5.2",
-        "@nimiq/keyguard-client": "^1.3.0",
+        "@nimiq/keyguard-client": "https://gitpkg.now.sh/nimiq/keyguard?529e33f5f0ad31d79e0f73e5073147a7e5e7fd58&scripts.postinstall=cd%20client%20%26%26%20.%2Fbuild-gitpkg.sh",
```
[the remark this answers is not in the document]
Comment (sisou) [reply package.json:25]: Fixed in `oasis` branch: c3aecd16b1e6895ff006eedc3f65c79aa87b41ba

File: src/lib/RpcApi.ts [hunk 4]
```diff
@@ -281,11 +287,15 @@ export default class RpcApi {
             if ((request as ParsedSimpleRequest).walletId) {
                 accountRequired = true;
                 account = await WalletStore.Instance.get((request as ParsedSimpleRequest).walletId);
-                if (!account) errorMsg = 'WalletId not found';
+                errorMsg = 'AccountId not found';
             } else if (requestType === RequestType.SIGN_TRANSACTION) {
                 accountRequired = true;
                 const address = (request as ParsedSignTransactionRequest).sender;
                 account = this._store.getters.findWalletByAddress(address.toUserFriendlyAddress(), true);
+            // } else if (requestType === RequestType.SIGN_BTC_TRANSACTION) {
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/RpcApi.ts:295]: Removed in #431: 57944671b7837926ed1937f74c42e0dd240e4eeb

File: src/lib/WalletInfoCollector.ts [hunk 5]
```diff
@@ -15,6 +15,20 @@ import {
...
 } from './LabelingMachine';
 import { VestingContractInfo } from './ContractInfo';
+import { BtcAddressInfo } from './bitcoin/BtcAddressInfo';
+import { loadBitcoinJS } from './bitcoin/BitcoinJSLoader';
+import { Receipt as BtcReceipt } from '@nimiq/electrum-client';
+import {
+    EXTERNAL_INDEX,
+    INTERNAL_INDEX,
+    BTC_ACCOUNT_MAX_ALLOWED_ADDRESS_GAP,
+    EXTENDED_KEY_PREFIXES,
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/WalletInfoCollector.ts:25]: Removed in #431: 16a118dab3d5d24038b6003a4c2281c640a739d7

File: src/lib/bitcoin/BitcoinConstants.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        MAIN: `m/49'/0'/0'`,
+        TEST: `m/49'/1'/0'`,
+    },
+    BIP84: {
+        MAIN: `m/84'/0'/0'`,
+        TEST: `m/84'/1'/0'`,
+    },
+};
+
+export const BTC_ACCOUNT_MAX_ALLOWED_ADDRESS_GAP = 5; // FIXME: Set to the standard 20 after testing
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/lib/bitcoin/BitcoinConstants.ts:24]: Both the Hub and Wallet currently use a gap of `10`. We need to evaluate performance and then update it in both coordinated. I have not yet heard of a case where 10 was not enough.

File: src/views/ActivateBitcoin.vue [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+                btcXPub: 'xpub6H1LXWLaKsWFhvm6RVpEL9P4KfRZSW7abD2ttkWP3SSQvnyA8FSVqNTEcYFgJS2UaFcxupHiYkro49S8yGasTvXEYBVPamhGW6cFJodrTHy',
+            };
+
+            if (!(await CookieHelper.canFitNewWallets([walletInfoEntry]))) {
+                this.notEnoughCookieSpace = true;
+                return;
+            }
+        }
+
+        const request: KeyguardClient.DeriveBtcXpubRequest = {
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/views/ActivateBitcoin.vue:45]: Fixed in #431: a363bbd4c6ce2f005281533ee239168dd5e4e487

File: src/views/ActivateBitcoinSuccess.vue [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import { WalletInfo } from '../lib/WalletInfo';
+import { WalletStore } from '../lib/WalletStore';
+import { Static } from '../lib/StaticStore';
+import StatusScreen from '../components/StatusScreen.vue';
+import WalletInfoCollector from '../lib/WalletInfoCollector';
+
+@Component({components: {StatusScreen, SmallPage}})
+export default class ActivateBitcoinSuccess extends Vue {
+    @Static private request!: ParsedSimpleRequest;
+    @State private keyguardResult!: KeyguardClient.DeriveBtcXpubResult;
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/views/ActivateBitcoinSuccess.vue:31]: Same as above.

File: src/views/ActivateBitcoin.vue [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            appName: this.request.appName,
+            keyId: walletInfo.keyId,
+            bitcoinXPubPath: BTC_ACCOUNT_KEY_PATH[Config.bitcoinAddressType][Config.bitcoinNetwork],
+        };
+
+        const client = this.$rpc.createKeyguardClient(true);
+        client.deriveBtcXpub(request);
```
[the remark this answers is not in the document]
Comment (sisou) [reply src/views/ActivateBitcoin.vue:52]: Same here.

File: src/views/ActivateBitcoinSuccess.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (sisou) [reply src/views/ActivateBitcoinSuccess.vue:59]: Removed in #431: 51aeec7ce1433c5edc7c7a1008356c2f927a8eb0

**Pull request merged**